### PR TITLE
Merging GOC and OSG email Scripts and updated them for tracking OSG 3.5 (SOFTWARE-3950)

### PR DESCRIPTION
**Pull request by Caffretro**
## 11/4/2020
- Updated email list tracking version to 3.5 in three scripts with `email-` prefix
- Formatted the usage of `goc-outdated-osg-pkgs` so the sequence of OSG version and EL version match the other two OSG scripts
- Added OSG 3.5 tracking to `osg-outdated-i2-pkgs`
- Merged `goc-outdated-osg-pkgs`, `osg-outdated-epel-pkgs`, `osg-outdated-i2-pkgs` into `outdated-pkgs`

### CODE REVIEW

**Comment by matyasselmeci** [comment]
Can you add el8 for the EPEL case while you're at it?

**Comment by Caffretro** [comment]
> Can you add el8 for the EPEL case while you're at it?

I guess you mean including EL8 for epel in the usage doc?

**Comment by matyasselmeci** [comment]
Sure, if it's only the usage docs that need updating.

File: outdated-pkgs [hunk 1]
```diff
@@ -0,0 +1,357 @@
+#!/usr/bin/python
+
+"""
+Usage:
+  For goc packages:
+      %s goc      [3.3|3.4|3.5] [el6|el7] [development|testing|release] [itb|production]
+  For osg epel packages:
+      %s osg-epel [3.4|3.5]     [el6|el7] [development|testing|release]
```
Comment (brianhlin) [inline outdated-pkgs:8]:
```suggestion
      %s osg-epel [3.4|3.5]     [el6|el7|el8] [development|testing|release]
```

File: outdated-pkgs [hunk 2]
```diff
@@ -0,0 +1,357 @@
+#!/usr/bin/python
+
+"""
+Usage:
+  For goc packages:
+      %s goc      [3.3|3.4|3.5] [el6|el7] [development|testing|release] [itb|production]
```
Comment (brianhlin) [inline outdated-pkgs:6]:
```suggestion
      %s goc      [3.3|3.4] [el6|el7] [development|testing|release] [itb|production]
```

File: outdated-pkgs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+  For osg epel packages:
+      %s osg-epel [3.4|3.5]     [el6|el7] [development|testing|release]
+    Or:
+      %s osg-epel [devops]      [el7]     [itb|production]
```
Comment (brianhlin) [inline outdated-pkgs:10]:
```suggestion
      %s osg-epel [devops]      [el7|el8]     [itb|production]
```

File: outdated-pkgs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+  For osg i2 packages:
+      %s osg-i2   [3.3|3.4|3.5] [el6|el7] [development|testing|release]
```
Comment (brianhlin) [inline outdated-pkgs:12]:
```suggestion
      %s osg-i2   [3.3|3.4] [el6|el7] [development|testing|release]
```

File: email-goc-outdated-osg-pkgs [hunk 5]
```diff
@@ -31,6 +31,6 @@ Subject: $subject
   } </dev/null &>/dev/null &
 }
 
-#check_for_el 6 3.4
-check_for_el 7 3.4
+#check_for_el 6 3.5
+check_for_el 7 3.5
```
Comment (brianhlin) [inline email-goc-outdated-osg-pkgs:35]: You can revert this change since we only have a `goc` repo for OSG 3.4

**Review by brianhlin — CHANGES_REQUESTED** [review]
I've made some suggestions to the various usages that should be reflected in the code if necessary.

Also, I'd prefer if you took one of the existing scripts (e.g. `osg-outdated-epel-pkgs`) and merged in changes from the other scripts, then renamed the whole thing to `outdated-pkgs`. That way, we'd retain some of the file history, which is useful when using `git blame` or looking at the `git log` for the single file.

**Review by edquist — COMMENTED** [review]
Not that i was assigned as a reviewer, but

> Also, I'd prefer if you took one of the existing scripts (e.g. osg-outdated-epel-pkgs) and merged in changes from the other scripts, then renamed the whole thing to outdated-pkgs. That way, we'd retain some of the file history, which is useful when using git blame or looking at the git log for the single file.

Yeah, note that for git blame to work right, you want to do the rename in a separate commit.  Either make your changes to osg-outdated-epel-pkgs, and then in a separate commit rename to outdated-pkgs; or else rename first (as a separate commit) and then make your changes to outdated-pkgs.

Otherwise, my serious grievance with this "merged" outdated-pkgs script so far is that it pastes in nearly-identical copies of the same code logic from the different scripts.  In particular the option parsing and the pkg diff calculation & output.

**Comment by Caffretro** [comment]
> Not that i was assigned as a reviewer, but
> 
> > Also, I'd prefer if you took one of the existing scripts (e.g. osg-outdated-epel-pkgs) and merged in changes from the other scripts, then renamed the whole thing to outdated-pkgs. That way, we'd retain some of the file history, which is useful when using git blame or looking at the git log for the single file.
> 
> Yeah, note that for git blame to work right, you want to do the rename in a separate commit. Either make your changes to osg-outdated-epel-pkgs, and then in a separate commit rename to outdated-pkgs; or else rename first (as a separate commit) and then make your changes to outdated-pkgs.
> 
> Otherwise, my serious grievance with this "merged" outdated-pkgs script so far is that it pastes in nearly-identical copies of the same code logic from the different scripts. In particular the option parsing and the pkg diff calculation & output.

Got it,

**Review by edquist — CHANGES_REQUESTED** [review]
Not sure if my previous comment got missed, but

>Otherwise, my serious grievance with this "merged" outdated-pkgs script so far is that it pastes in nearly-identical copies of the same code logic from the different scripts. In particular the option parsing and the pkg diff calculation & output.

File: outdated-pkgs [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        pkg_diffs[:0] = [["Package", "OSG " + SERIES, "I2 EL" + EL]]
+        widths = [ max(map(len,col)) for col in zip(*pkg_diffs) ]
+        pkg_diffs[1:1] = [[ '-' * n for n in widths ]]
+        for i,row in enumerate(pkg_diffs):
+            spacing = [ w-len(x) for x,w in zip(row,widths) ]
+            if use_color and i > 1:
+                row[1:] = colorize_vr(*row[1:])
+            print '  '.join( r + ' ' * s for r,s in zip(row,spacing) ).rstrip()
+    else:
+        print "No package version differences"
```
Comment (brianhlin) [inline outdated-pkgs:356]: Yeah as @edquist is saying, this looks like a lot of duplicated logic that can be refactored. Similarly for the argparsing above, e.g. each of the "modes" has the following

```
        elif arg == '--show-all':
            show_all = True
        else:
            usage()
```

So you could take them out of the mode `if` blocks and just check them once.

Comment (Caffretro) [reply outdated-pkgs]: I get the idea now. Just pushed two commits that generalized repeated functions.

File: outdated-pkgs [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+    goc_map = nvrmap(get_koji_list(goc_tag))
+    osg_map = nvrmap(get_koji_list(osg_tag))
+
+    get_pkg_diff(goc_map, osg_map)
+    handle_pkg_diff()
+elif MODE == 'osg-epel':
+    epel_map = nvrmap(get_epel_list())
+    osg_map  = nvrmap(get_osg_list())
+    get_pkg_diff(epel_map, osg_map)
```
Comment (edquist) [inline outdated-pkgs:296]: I think you have these backwards

```suggestion
    get_pkg_diff(osg_map, epel_map)
```

File: outdated-pkgs [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    handle_pkg_diff()
+elif MODE == 'osg-i2':
+    i2_map = nvrmap(get_i2_list())
+    osg_map  = nvrmap(get_osg_list())
+    get_pkg_diff(i2_map, osg_map)
```
Comment (edquist) [inline outdated-pkgs:301]:
```suggestion
    get_pkg_diff(osg_map, i2_map)
```

File: outdated-pkgs [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            pkg_diffs[:0] = [["Package", "OSG " + SERIES, "I2 EL" + EL]]
+        widths = [ max(map(len,col)) for col in zip(*pkg_diffs) ]
+        pkg_diffs[1:1] = [[ '-' * n for n in widths ]]
+        for i,row in enumerate(pkg_diffs):
+            spacing = [ w-len(x) for x,w in zip(row,widths) ]
+            if use_color and i > 1:
+                row[1:] = colorize_vr(*row[1:])
+            print '  '.join( r + ' ' * s for r,s in zip(row,spacing) ).rstrip()
+    else:
+        print "No package version differences"
```
Comment (edquist) [inline outdated-pkgs:282]: This is a good start for breaking out the common functionality.

I like your idea of making separate functions for `get_pkg_diff()` and `handle_pkg_diff()` - since each block of code has its own discrete logical task.  But if you are going to separate them this way, i would suggest that `pkg_diffs` can be returned by `get_pkg_diff()` and taken as a parameter to `handle_pkg_diff()`, rather than making it a global.  (Though it should work as you have it.)

File: outdated-pkgs [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+                pkg_diffs.append([pkg, source[pkg], compare[pkg]])
+
+def handle_pkg_diff():
+    if pkg_diffs:
+        if MODE == 'goc':
+            pkg_diffs[:0] = [["Package", "EL%s GOC" % EL, "OSG " + SERIES]]
+        elif MODE == 'osg-epel':
+            pkg_diffs[:0] = [["Package", "OSG " + SERIES, "EPEL " + EL]]
+        elif MODE == 'osg-i2':
+            pkg_diffs[:0] = [["Package", "OSG " + SERIES, "I2 EL" + EL]]
```
Comment (edquist) [inline outdated-pkgs:273]: For consideration / extra credit:

It seems to me that the repo labels here could be passed in as parameters to this function rather than having the repeated if/elif/elif switch based on MODE.

File: outdated-pkgs [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+        elif re.search(r'^(upcoming-)?(development|testing|release)$', arg) \
+            or arg in ('itb', 'production'):
+            TAG = arg
+    elif MODE == 'osg-i2':
+        if re.search(r'^el[5-9]$', arg):
+            EL = arg[2]
+        elif re.search(r'^[3-9]\.[0-9]$', arg):
+            SERIES = arg
+        elif re.search(r'^(upcoming-)?(development|testing|release)$', arg):
+            TAG = arg
```
Comment (edquist) [inline outdated-pkgs:98]:
```suggestion
    if re.search(r'^el[5-9]$', arg):
        EL = arg[2]
    elif re.search(r'^[3-9]\.[0-9]$', arg) or arg == 'devops':
        SERIES = arg
    elif re.search(r'^(upcoming-)?(development|testing|release)$', arg) \ 
        or arg in ('itb', 'production'):
        TAG = arg
```

Seems like all this can be collapsed, and then you can change existing `OSG_REPO` references to `TAG`

File: outdated-pkgs [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+    # check which funcion user used, print different values
+    script = os.path.basename(__file__)
+    print __doc__ % (script, script, script, script,
+                    SERIES, EL, OSG_REPO, GOC_REPO,
+                    SERIES, EL, TAG,
+                    SERIES, EL, TAG)
+    sys.exit()
+
+# Mode selection
+if sys.argv[1] in ['goc', 'osg-epel', 'osg-i2']:
```
Comment (edquist) [inline outdated-pkgs:64]: For consistency, i would suggest we call the `goc` mode `goc-osg`, to follow the pattern that it shows goc packages that are behind their osg counterparts.

(Although i don't know if we even care about the goc-outdated-osg-pkgs script anymore, so maybe all the goc mode stuff will be removed?)

```suggestion
if sys.argv[1] in ['goc-osg', 'osg-epel', 'osg-i2']:
```

File: outdated-pkgs (hunk 7 above)
Comment (Caffretro) [reply outdated-pkgs:296]: you're right

File: outdated-pkgs (hunk 9 above)
Comment (Caffretro) [reply outdated-pkgs:282]: I guess you mean move the handle function under the other?

Comment (Caffretro) [reply outdated-pkgs:282]: Nevermind I get what you mean. I was trying to keep the original logic since I'm not sure if there are any special considerations behind setting that list to global.

File: outdated-pkgs [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+  For goc packages:
+      %s goc-osg  [3.3|3.4]     [el6|el7] [development|testing|release] [itb|production]
+  For osg epel packages:
+      %s osg-epel [3.4|3.5|3.5] [el6|el7] [development|testing|release]
+    Or:
+      %s osg-epel [devops]      [el7|el8] [itb|production]
+  For osg i2 packages:
+      %s osg-i2   [3.3|3.4]     [el6|el7] [development|testing|release]
+
+Default for goc:
```
Comment (edquist) [inline outdated-pkgs:14]: With the combined script, the usage summary is less obvious what the different modes are actually doing.  I feel like a few more words can help to clarify:

```suggestion
  For goc packages behind osg versions:
      %s goc-osg  [3.3|3.4]     [el6|el7] [development|testing|release] [itb|production]
  For osg packages behind epel versions:
      %s osg-epel [3.4|3.5|3.5] [el6|el7] [development|testing|release]
    Or:
      %s osg-epel [devops]      [el7|el8] [itb|production]
  For osg packages behind i2 versions:
      %s osg-i2   [3.3|3.4]     [el6|el7] [development|testing|release]

Default for goc-osg:
```

**Review by edquist — COMMENTED** [review]
Thanks @Caffretro,

I hope this was an enjoyable exercise for you, to find a way to factor out the common logical parts of these similar tasks.

If you'd like, you can verify whether the output of your new script (for various combinations of parameters) matches the output of the old scripts in master.

I also made a suggestion to make the usage summary (hopefully) a bit more clear.

Beyond that i don't have any more requests for you at the moment.

I made some comments for @matyasselmeci & @brianhlin in SOFTWARE-3950 though. Sorry to wait till now to ask, but i would like to confirm that we actually still care about the goc and i2 functionality before merging this in.

**Review by brianhlin — CHANGES_REQUESTED** [review]

File: outdated-pkgs [hunk 14]
```diff
@@ -0,0 +1,290 @@
+#!/usr/bin/python
+
+"""
+Usage:
+  For goc packages behind osg versions:
+      %s goc-osg  [3.3|3.4]     [el6|el7] [development|testing|release] [itb|production]
```
[the remark this answers is not in the document]
Comment (edquist) [reply outdated-pkgs:6]: Actually, as i tried to bring up in SOFTWARE-3950; for better or for worse it seems the only functionality we still care about is the original osg-outdated-epel-pkgs script.  Not that we don't appreciate the work that went into merging the others in, but in the end those parts won't be used.

**Comment by Caffretro** [comment]
I kept the function `get_koji_list` used for the GOC repo and a lot of if statements related to `MODE ` identification in case the script will be developed with more features.

**Comment by brianhlin** [comment]
We decided to abandon this project
